File: hooks/rules/tentacle.py
"""Tentacle enforcement + suggestion (merged).

Combines enforce-tentacle.py and tentacle-suggest.py into one module.
Single get_module() implementation, shared edit tracking.

Edit tracking uses a per-repo-partitioned JSON payload so edits in one git
repository never inflate counts for a different repository.  Entries carry a
Unix timestamp and are dropped after 24 h (TTL).
"""
import json
import re
import subprocess
import sys
import time
from pathlib import Path

from . import Rule
from .common import MARKERS_DIR, CODE_EXTENSIONS, get_module, is_session_path, is_source_path, deny, info

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
try:
    from marker_auth import (verify_marker, verify_list_marker, sign_list_marker,
                             is_secret_access, check_tamper_marker)
except ImportError:
    def verify_marker(p, n): return False
    def verify_list_marker(p): return set()
    def sign_list_marker(p, lines):
        p.parent.mkdir(parents=True, exist_ok=True)
        p.write_text("\n".join(sorted(lines)))
    def is_secret_access(c): return True
    def check_tamper_marker(): return False

EDITS_FILE = MARKERS_DIR / "tentacle-edits"
TENTACLE_DONE = MARKERS_DIR / "tentacle-done"
TENTACLE_BYPASS = MARKERS_DIR / "tentacle-bypass"
SUGGESTED_FILE = MARKERS_DIR / "tentacle-suggested"

MIN_FILES = 3
MIN_MODULES = 2


# ── Edit-tracking helpers ──────────────────────────────────────────────────

def _get_git_root(cwd=None):
    """Return the git root for *cwd* (or cwd()) as a string, or None."""
    try:
        r = subprocess.run(
            ["git", "-C", str(cwd or Path.cwd()), "rev-parse", "--show-toplevel"],
            capture_output=True, text=True, timeout=5,
        )
        if r.returncode == 0:
            return r.stdout.strip()
    except Exception:
        pass
    return None


def _prune_ttl(entries, now):
    """Return entries whose timestamp is within the last 24 hours."""
    cutoff = now - 86400
    return [e for e in entries if isinstance(e, dict) and e.get("t", 0) >= cutoff]


def _read_edits(path):
    """Read the HMAC-signed edits marker and return a dict of the form::

        {git_root: [{"p": file_path, "t": unix_ts}, ...], ...}

    Accepts both:
    * **New format** – the signed content is a single-element set whose sole
      element is a JSON-encoded dict (written by ``_write_edits``).
    * **Legacy format** – the signed content is a set of bare file-path
      strings (old format).  They are migrated into a ``"legacy"`` bucket
      with the current timestamp so they expire naturally after 24 h.
    """
    raw_set = verify_list_marker(path)
    if not raw_set:
        return {}
    # New format: exactly one element that is a JSON dict
    if len(raw_set) == 1:
        sole = next(iter(raw_set))
        if sole.startswith("{"):
            try:
                data = json.loads(sole)
                if isinstance(data, dict):
                    return data
            except (json.JSONDecodeError, ValueError):
                pass
    # Legacy: flat set of file paths → migrate with current timestamp
    now = time.time()
    legacy_entries = [{"p": fp, "t": now} for fp in raw_set if fp]
    return {"legacy": legacy_entries}


def _write_edits(path, data):
    """Write *data* as an HMAC-signed single-JSON-element list marker."""
    payload = json.dumps(data, separators=(",", ":"), sort_keys=True)
    sign_list_marker(path, {payload})


def _get_entries_for_repo(data, git_root):
    """Return entries that belong to *git_root*.

    * If *git_root* is present as a key, return those entries directly.
    * Otherwise fall back to the ``"legacy"`` bucket, filtering by path
      prefix when *git_root* is known (heuristic for migrated markers).
    * If *git_root* is ``None`` (git unavailable), return all legacy entries
      without filtering.
    """
    if git_root and git_root in data:
        return list(data[git_root])
    if "legacy" in data:
        if git_root:
            return [e for e in data["legacy"] if str(e.get("p", "")).startswith(git_root)]
        return list(data["legacy"])
    return []


class TentacleEnforceRule(Rule):
    """Block edits across too many modules without tentacle-orchestration."""

    name = "tentacle-enforce"
    events = ["preToolUse"]
    tools = ["edit", "create", "bash"]

    def evaluate(self, event, data):
        tool_name = data.get("toolName", "")
        tool_args = data.get("toolArgs", {})
        if not isinstance(tool_args, dict):
            tool_args = {}

        # Kill-switch
        if check_tamper_marker():
            return deny(
                "\U0001f6a8 HOOKS TAMPERED: All modifications blocked. "
                "Run: sudo python3 ~/.copilot/tools/install.py --lock-hooks"
            )

        # For bash, check if it writes source files
        if tool_name == "bash":
            command = tool_args.get("command", "")
            if is_secret_access(command):
                return deny("\U0001f512 Access to protected hook files is blocked.")
            source_exts = tuple(CODE_EXTENSIONS)
            writes_source = False
            if any(ext in command for ext in source_exts):
                if any(p in command for p in ("<<", "write_text", "open(",
                                               "sed -i", "tee ", "cp ", "mv ",
                                               "dd ", "patch ", "rsync ", "install ")):
                    writes_source = True
                elif re.search(r">{1,2}\s*\S+", command):
                    for m in re.finditer(r">{1,2}\s*([^\s;|&]+)", command):
                        p = m.group(1)
                        # Strip surrounding shell quotes before path checks
                        if len(p) >= 2 and p[0] == p[-1] and p[0] in ('"', "'"):
                            p = p[1:-1]
                        if is_source_path(p):
                            writes_source = True
                            break
            if not writes_source:
                return None

        # FP-1: session-state files (e.g. /research outputs) are not project source;
        # skip threshold check so create/edit to these paths is never blocked.
        if tool_name in ("edit", "create"):
            file_path = tool_args.get("path", "")
            if file_path and is_session_path(file_path):
                return None

        # Check bypass markers
        if verify_marker(TENTACLE_DONE, "tentacle-done"):
            return None
        if verify_marker(TENTACLE_BYPASS, "tentacle-bypass"):
            return None

        # Check tracked edits — only count files for the current repo
        git_root = _get_git_root()
        repo_prefix = Path(git_root).name if git_root else "legacy"
        edited_dict = _read_edits(EDITS_FILE)
        now = time.time()
        for key in list(edited_dict.keys()):
            edited_dict[key] = _prune_ttl(edited_dict[key], now)
        entries = _get_entries_for_repo(edited_dict, git_root)
        if not entries or len(entries) < MIN_FILES:
            return None

        modules = {get_module(e["p"], repo_prefix) for e in entries if get_module(e["p"])}
        if len(modules) < MIN_MODULES:
            return None

        return deny(
            f"\U0001f419 TENTACLE REQUIRED: {len(entries)} files across {len(modules)} modules "
            f"({', '.join(sorted(modules))}). "
            "Multi-module edits should use tentacle-orchestration. "
            "If you are the orchestrator: (1) tentacle.py create <name> --scope \"<paths>\" --desc \"<desc>\" --briefing  "
            "(2) tentacle.py todo <name> add \"<task>\"  "
            "(3) tentacle.py swarm <name> --agent-type general-purpose --model claude-sonnet-4.6  "
            "If you are a dispatched sub-agent: stay within your assigned scope, write results to "
            "handoff.md, and avoid git commit or git push — by convention the orchestrator "
            "commits and pushes after all tentacles are verified."
        )


class TentacleSuggestRule(Rule):
    """Suggest tentacle when edits span multiple modules (postToolUse)."""

    name = "tentacle-suggest"
    events = ["postToolUse"]
    tools = ["edit", "create", "bash"]

    def evaluate(self, event, data):
        tool_name = data.get("toolName", "")
        tool_args = data.get("toolArgs", {})
        if not isinstance(tool_args, dict):
            tool_args = {}

        if SUGGESTED_FILE.is_file():
            return None

        # Collect edited file paths
        file_paths = []
        if tool_name in ("edit", "create"):
            fp = ""
            if tool_name == "edit":
                fp = (data.get("toolResult") or {}).get("filePath", "")
            elif tool_name == "create":
                fp = (data.get("input") or {}).get("filePath", "")
            if fp:
                file_paths.append(fp)
        elif tool_name == "bash":
            command = tool_args.get("command", "")
            if "<<" in command and "open(" in command:
                for m in re.finditer(r"open\(['\"]([^'\"]+)['\"]", command):
                    p = m.group(1)
                    if not p.startswith(("/tmp/", "/var/", "/dev/")):
                        file_paths.append(p)
            if ">" in command:
                for m in re.finditer(r">{1,2}\s*([^\s;|&]+)", command):
                    p = m.group(1)
                    # Strip surrounding shell quotes before path checks
                    if len(p) >= 2 and p[0] == p[-1] and p[0] in ('"', "'"):
                        p = p[1:-1]
                    if not p.startswith(("/tmp/", "/var/", "/dev/")):
                        file_paths.append(p)
            # Mirror enforce: also track paths written by sed -i and tee
            for m in re.finditer(r"\bsed\s+-i[^\s]*\s+(?:'[^']*'|\"[^\"]*\")\s+(\S+)", command):
                p = m.group(1)
                if not p.startswith(("/tmp/", "/var/", "/dev/")):
                    file_paths.append(p)
            for m in re.finditer(r"\btee\s+(?:-[a-z]+\s+)?(\S+)", command):
                p = m.group(1)
                if not p.startswith(("/tmp/", "/var/", "/dev/")):
                    file_paths.append(p)

        if not file_paths:
            return None

        # Filter: only track actual code/config files, not markdown or session-state paths.
        # This prevents session-research markdown writes from accumulating in the
        # tentacle-edits marker and falsely triggering multi-module enforcement.
        file_paths = [
            fp for fp in file_paths
            if Path(fp).suffix.lower() in CODE_EXTENSIONS and not is_session_path(fp)
        ]

        if not file_paths:
            return None

        # Track edited files using HMAC-signed list markers, partitioned by git repo
        MARKERS_DIR.mkdir(parents=True, exist_ok=True)
        git_root = _get_git_root()
        repo_key = git_root if git_root else "legacy"
        repo_prefix = Path(git_root).name if git_root else "legacy"
        now = time.time()
        edited_dict = _read_edits(EDITS_FILE)
        # Prune TTL on every bucket
        for key in list(edited_dict.keys()):
            edited_dict[key] = _prune_ttl(edited_dict[key], now)
        entries = edited_dict.get(repo_key, [])
        existing_paths = {e["p"] for e in entries}
        for fp in file_paths:
            if fp not in existing_paths:
                entries.append({"p": fp, "t": now})
                existing_paths.add(fp)
        edited_dict[repo_key] = entries
        try:
            _write_edits(EDITS_FILE, edited_dict)
        except Exception:
            pass

        if len(entries) < MIN_FILES:
            return None

        modules = {get_module(e["p"], repo_prefix) for e in entries if get_module(e["p"])}
        if len(modules) < MIN_MODULES:
            return None

        try:
            SUGGESTED_FILE.touch()
        except Exception:
            pass

        return info(
            f"\n  \U0001f419 TENTACLE SUGGESTION: {len(entries)} files across "
            f"{len(modules)} modules detected.\n"
            "  Consider using tentacle-orchestration for parallel multi-agent execution.\n"
            f"  Modules: {', '.join(sorted(modules))}\n"
        )

File: sk-rust/src/hooks/rules/tentacle.rs
use super::*;

// ---------------------------------------------------------------------------
// Shared helpers for VerificationGatePreRule + TentacleSuggestRule
// ---------------------------------------------------------------------------

/// Module-marker directory names — mirrors Python `MODULE_MARKERS` tuple
/// in `hooks/rules/common.py`.
pub(crate) const MODULE_MARKERS: &[&str] = &[
    "src",
    "lib",
    "app",
    "pkg",
    "internal",
    "cmd",
    "hooks",
    "skills",
    "templates",
    "tests",
    "test",
    "components",
    "screens",
    "services",
    "utils",
    "models",
    "views",
    "controllers",
    "routes",
    "pages",
    "features",
    "presentation",
    "domain",
    "data",
    "core",
    "common",
    "ui",
    "api",
    "db",
    "auth",
    "config",
    "settings",
    "alarm",
    "timer",
    "stopwatch",
    "clock",
    "widget",
];

/// Extract the deepest module name from a file path.
///
/// Mirrors Python `get_module(file_path, repo_prefix)` in
/// `hooks/rules/common.py`.  Returns an empty string when no module can be
/// determined.  Normalises `\` to `/` so Windows paths work correctly.
pub(crate) fn get_module_for_path(file_path: &str, repo_prefix: Option<&str>) -> String {
    let norm = file_path.replace('\\', "/");
    let parts: Vec<&str> = norm.split('/').filter(|s| !s.is_empty()).collect();
    if parts.len() < 2 {
        return String::new();
    }
    // Iterate all parts except the last (filename).
    let mut best = String::new();
    for i in 0..parts.len() - 1 {
        let p = parts[i];
        if MODULE_MARKERS.contains(&p) {
            // i+1 < parts.len()-1 means the next element is not the filename.
            if i + 1 < parts.len() - 1 {
                best = format!("{}/{}", p, parts[i + 1]);
            } else {
                best = p.to_string();
            }
        }
    }
    let module = if !best.is_empty() {
        best
    } else {
        // Fallback: parent directory name.
        parts[parts.len() - 2].to_string()
    };
    if module.is_empty() {
        return String::new();
    }
    match repo_prefix {
        Some(prefix) if !prefix.is_empty() => format!("{}:{}", prefix, module),
        _ => module,
    }
}

/// Run `git rev-parse --show-toplevel` and return the absolute path string.
///
/// Returns `None` on any error (fail-open: git not in PATH, not a repo).
pub(crate) fn get_git_root() -> Option<String> {
    let output = Command::new("git")
        .args(["rev-parse", "--show-toplevel"])
        .stdout(Stdio::piped())
        .stderr(Stdio::null())
        .output()
        .ok()?;
    if !output.status.success() {
        return None;
    }
    let s = String::from_utf8_lossy(&output.stdout).trim().to_string();
    if s.is_empty() {
        None
    } else {
        Some(s)
    }
}

/// Return the short basename of the git root (repo prefix for module names),
/// or `"legacy"` when git is unavailable.
pub(crate) fn get_repo_prefix() -> String {
    get_git_root()
        .and_then(|root| {
            std::path::Path::new(&root)
                .file_name()
                .and_then(|n| n.to_str())
                .map(|s| s.to_string())
        })
        .unwrap_or_else(|| "legacy".to_string())
}

// ---------------------------------------------------------------------------
// TentacleSuggestRule
// ---------------------------------------------------------------------------

/// Suggest tentacle-orchestration on postToolUse when edits span multiple modules.
///
/// Informational-only port of
/// `hooks/rules/tentacle.py::TentacleSuggestRule` (wave8):
///   - On postToolUse [edit/create/bash]: reads the `tentacle-edits` HMAC-signed
///     list marker (written by `TrackEditsRule`) and counts distinct files and
///     modules touched so far.
///   - If `>= 3` files AND `>= 2` modules are detected: emits a suggestion,
///     then touches `tentacle-suggested` so the suggestion fires only once.
///   - Skips immediately when `tentacle-suggested` is already present.
///
/// Read-only with respect to `tentacle-edits`:
///   - This rule NEVER writes to `tentacle-edits`.
///   - `TrackEditsRule` is the sole writer; `TentacleSuggestRule` only reads.
///
/// Format compatibility:
///   - Reads both the legacy flat-path format (written by Rust `TrackEditsRule`)
///     and the new JSON-dict format (written by Python `TentacleSuggestRule`).
///   - Legacy: each element in the HMAC-signed set is a bare file path string.
///   - New: a single element is a JSON dict `{repo_root: [{p, t}, ...], ...}`.
///
/// Informational-only: never returns a deny result.  Fail-open at every step.
pub struct TentacleSuggestRule;

pub(crate) const SUGGEST_MIN_FILES: usize = 3;
pub(crate) const SUGGEST_MIN_MODULES: usize = 2;

/// Read file paths from the `tentacle-edits` HMAC-signed list marker.
///
/// Handles both:
/// - Legacy flat format: each element is a bare file-path string (Rust writer).
/// - New JSON-dict format: single element is `{repo_root: [{p, t}...], ...}`
///   (Python writer).
///
/// TTL pruning is not applied to the legacy flat format (no timestamps stored).
/// Fail-open: any parse error → returns empty vec.
pub(crate) fn read_tentacle_edits_paths() -> Vec<String> {
    let edits_path = markers_dir().join("tentacle-edits");
    let raw_set = marker_auth::verify_list_marker(&edits_path);
    let mut paths = Vec::new();
    for entry in &raw_set {
        if entry.starts_with('{') {
            // New JSON-dict format: {repo_root: [{p: path, t: ts}, ...], ...}
            if let Ok(val) = serde_json::from_str::<Value>(entry) {
                if let Some(obj) = val.as_object() {
                    for (_, bucket) in obj {
                        if let Some(arr) = bucket.as_array() {
                            for e in arr {
                                if let Some(p) = e.get("p").and_then(|v| v.as_str()) {
                                    paths.push(p.to_string());
                                }
                            }
                        }
                    }
                }
            }
        } else if !entry.is_empty() {
            // Legacy flat format: bare file path.
            paths.push(entry.clone());
        }
    }
    paths
}

impl HookRule for TentacleSuggestRule {
    fn name(&self) -> &'static str {
        "tentacle-suggest"
    }

    fn events(&self) -> &'static [&'static str] {
        &["postToolUse"]
    }

    fn tools(&self) -> &'static [&'static str] {
        &["edit", "create", "bash"]
    }

    fn evaluate(&self, _event: &str, data: &Value) -> Option<Value> {
        let tool_name = data.get("toolName").and_then(|v| v.as_str()).unwrap_or("");
        if tool_name != "edit" && tool_name != "create" && tool_name != "bash" {
            return None;
        }

        let suggested_path = markers_dir().join("tentacle-suggested");
        if suggested_path.is_file() {
            return None; // suggestion already emitted this session
        }

        // Read current tentacle-edits state (written by TrackEditsRule).
        let all_paths = read_tentacle_edits_paths();
        if all_paths.len() < SUGGEST_MIN_FILES {
            return None;
        }

        // Deduplicate paths (the marker may contain duplicates after migration).
        let unique_paths: std::collections::HashSet<String> = all_paths.into_iter().collect();
        if unique_paths.len() < SUGGEST_MIN_FILES {
            return None;
        }

        // Filter to code/config extensions only (no markdown, no session-state).
        let code_paths: Vec<&str> = unique_paths
            .iter()
            .map(|s| s.as_str())
            .filter(|p| has_code_extension(p) && !is_track_session_path(p))
            .collect();
        if code_paths.len() < SUGGEST_MIN_FILES {
            return None;
        }

        // Compute modules (fail-open: git unavailable → use "legacy" prefix).
        let repo_prefix = get_repo_prefix();
        let modules: std::collections::HashSet<String> = code_paths
            .iter()
            .map(|p| get_module_for_path(p, Some(&repo_prefix)))
            .filter(|m| !m.is_empty())
            .collect();

        if modules.len() < SUGGEST_MIN_MODULES {
            return None;
        }

        // Touch tentacle-suggested so this fires only once.
        let mdir = markers_dir();
        let _ = fs::create_dir_all(&mdir);
        let _ = fs::write(&suggested_path, b"");

        let mut sorted_modules: Vec<&str> = modules.iter().map(|s| s.as_str()).collect();
        sorted_modules.sort_unstable();

        Some(info(&format!(
            "\n  \u{1f419} TENTACLE SUGGESTION: {} files across {} modules detected.\n  \
             Consider using tentacle-orchestration for parallel multi-agent execution.\n  \
             Modules: {}\n  \
             \u{1f4cb} After completing: check docs/SYNC-MATRIX.md for docs/memory follow-ups.\n",
            code_paths.len(),
            modules.len(),
            sorted_modules.join(", "),
        )))
    }
}

// ---------------------------------------------------------------------------
// TentacleEnforceRule — wave12
// ---------------------------------------------------------------------------

/// Minimum file count before `TentacleEnforceRule` fires.
pub(crate) const TENTACLE_ENFORCE_MIN_FILES: usize = 3;
/// Minimum module count before `TentacleEnforceRule` fires.
pub(crate) const TENTACLE_ENFORCE_MIN_MODULES: usize = 2;
/// TTL for `tentacle-edits` entries (24 hours in seconds).
pub(crate) const TENTACLE_ENFORCE_TTL_SECS: u64 = 86_400;

/// Read tentacle-edits paths for the current git repo, applying per-repo
/// filtering and 24-hour TTL pruning.
///
/// Handles both:
/// - **New JSON-dict format** (`{repo_root: [{p, t}…], …}`): filters to the
///   bucket matching the current git root; falls back to the "legacy" bucket
///   with path-prefix filtering.
/// - **Legacy flat format** (bare file-path strings written by
///   `TrackEditsRule`): collected into a synthetic "legacy" bucket with
///   the current timestamp so they expire naturally after 24 h, mirroring
///   Python `_read_edits()`. When a git root is available, relative legacy
///   paths are treated as belonging to the current repo so the native direct
///   writer (`TrackEditsRule`) remains compatible.
///
/// Mirrors `TentacleEnforceRule._read_edits()` and
/// `TentacleEnforceRule._get_entries_for_repo()` in
/// `hooks/rules/tentacle.py`.
pub(crate) fn read_tentacle_edits_for_current_repo() -> Vec<String> {
    let edits_path = markers_dir().join("tentacle-edits");
    let raw_set = marker_auth::verify_list_marker(&edits_path);
    if raw_set.is_empty() {
        return Vec::new();
    }

    let git_root = get_git_root();
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs();
    let ttl_cutoff = now.saturating_sub(TENTACLE_ENFORCE_TTL_SECS);

    // Parse HMAC-set into repo_root → Vec<(path, timestamp)> buckets.
    let mut repo_buckets: std::collections::HashMap<String, Vec<(String, u64)>> =
        std::collections::HashMap::new();
    for entry in &raw_set {
        if entry.starts_with('{') {
            // New JSON-dict format: {repo_root: [{p, t}, ...], ...}
            if let Ok(val) = serde_json::from_str::<Value>(entry) {
                if let Some(obj) = val.as_object() {
                    for (key, bucket) in obj {
                        if let Some(arr) = bucket.as_array() {
                            let vec = repo_buckets.entry(key.clone()).or_default();
                            for e in arr {
                                let p = match e.get("p").and_then(|v| v.as_str()) {
                                    Some(s) => s.to_string(),
                                    None => continue,
                                };
                                let t = e.get("t").and_then(|v| v.as_f64()).unwrap_or(0.0) as u64;
                                vec.push((p, t));
                            }
                        }
                    }
                }
            }
        } else if !entry.is_empty() {
            // Legacy flat path: migrate with current timestamp so TTL still applies.
            repo_buckets
                .entry("legacy".to_string())
                .or_default()
                .push((entry.clone(), now));
        }
    }

    // Select and filter entries for the current repo.
    let root_ref = git_root.as_deref();
    let entries: Vec<(String, u64)> = if let Some(root) = root_ref {
        if let Some(bucket) = repo_buckets.get(root) {
            bucket.clone()
        } else if let Some(legacy) = repo_buckets.get("legacy") {
            // Direct-path compatibility: legacy entries written by native
            // TrackEditsRule are relative paths, so treat relative legacy
            // entries as belonging to the current repo. Absolute paths still
            // require a git-root prefix match.
            legacy
                .iter()
                .filter(|(p, _)| p.starts_with(root) || !Path::new(p).is_absolute())
                .cloned()
                .collect()
        } else {
            Vec::new()
        }
    } else {
        // Python parity: when git root is unavailable, only use the legacy bucket.
        repo_buckets.get("legacy").cloned().unwrap_or_default()
    };

    // Apply TTL: keep entries within the last 24 hours.
    entries
        .into_iter()
        .filter(|(_, t)| *t >= ttl_cutoff)
        .map(|(p, _)| p)
        .collect()
}

/// Return `true` when the bash command appears to write a source file,
/// using `TentacleEnforceRule` semantics.
///
/// Extends the standard redirect/heredoc/sed detection with additional
/// destructive-write indicators: `cp`, `mv`, `patch`, `rsync`, `install`.
/// Mirrors the inline bash-analysis in `TentacleEnforceRule.evaluate()` in
/// `hooks/rules/tentacle.py`.
pub(crate) fn bash_writes_source_for_enforce_tentacle(command: &str) -> bool {
    // Quick exit: no code extension present in the command.
    if !TRACK_CODE_EXTENSIONS
        .iter()
        .any(|ext| command.contains(ext))
        && !ENFORCE_SOURCE_EXTENSIONS
            .iter()
            .any(|ext| command.contains(ext))
    {
        return false;
    }

    // Explicit write-pattern indicators (mirrors Python TentacleEnforceRule).
    if command.contains("<<")
        || command.contains("write_text")
        || command.contains("open(")
        || command.contains("sed -i")
        || command.contains("tee ")
        || command.contains("cp ")
        || command.contains("mv ")
        || command.contains("dd ")
        || command.contains("patch ")
        || command.contains("rsync ")
        || command.contains("install ")
    {
        return true;
    }

    // Redirect `>` / `>>` → check whether each target is a source file.
    let written = extract_written_paths_simple(command);
    for path in &written {
        if is_source_path_for_enforce(path) {
            return true;
        }
    }

    false
}

/// Deny multi-module edits that should go through tentacle-orchestration.
///
/// Mirrors `TentacleEnforceRule` in `hooks/rules/tentacle.py`.
///
/// Triggers on `preToolUse` for `edit`/`create`/`bash` when the
/// `tentacle-edits` marker shows ≥ 3 files across ≥ 2 modules without a
/// valid `tentacle-done` or `tentacle-bypass` marker.  Applies per-repo
/// filtering and 24-hour TTL pruning on JSON-dict entries.
pub struct TentacleEnforceRule;

impl HookRule for TentacleEnforceRule {
    fn name(&self) -> &'static str {
        "tentacle-enforce"
    }

    fn events(&self) -> &'static [&'static str] {
        &["preToolUse"]
    }

    fn tools(&self) -> &'static [&'static str] {
        &["edit", "create", "bash"]
    }

    fn evaluate(&self, _event: &str, data: &Value) -> Option<Value> {
        let tool_name = data.get("toolName").and_then(|v| v.as_str()).unwrap_or("");
        let tool_args = data
            .get("toolArgs")
            .and_then(|v| v.as_object())
            .cloned()
            .unwrap_or_default();

        // Kill-switch: hooks-tampered → deny all modifications.
        if marker_auth::check_tamper_marker() {
            return Some(deny(
                "\u{1f6a8} HOOKS TAMPERED: All modifications blocked. Run: sudo python3 ~/.copilot/tools/install.py --lock-hooks",
            ));
        }

        // Bash: secret-access check first, then source-write detection.
        if tool_name == "bash" {
            let command = tool_args
                .get("command")
                .and_then(|v| v.as_str())
                .unwrap_or("");
            if marker_auth::is_secret_access(command) {
                return Some(deny("\u{1f512} Access to protected hook files is blocked."));
            }
            if !bash_writes_source_for_enforce_tentacle(command) {
                return None;
            }
        }

        // Edit / create: skip session-state paths (false-positive bypass).
        if tool_name == "edit" || tool_name == "create" {
            let file_path = tool_args.get("path").and_then(|v| v.as_str()).unwrap_or("");
            if !file_path.is_empty() && is_track_session_path(file_path) {
                return None;
            }
        }

        // Bypass markers: tentacle-done or tentacle-bypass.
        let mdir = markers_dir();
        if marker_auth::verify_marker(&mdir.join("tentacle-done"), "tentacle-done") {
            return None;
        }
        if marker_auth::verify_marker(&mdir.join("tentacle-bypass"), "tentacle-bypass") {
            return None;
        }

        // Read tentacle-edits with per-repo filtering and 24-hour TTL pruning.
        let paths = read_tentacle_edits_for_current_repo();
        if paths.len() < TENTACLE_ENFORCE_MIN_FILES {
            return None;
        }

        // Compute unique modules; deny when ≥ 2 distinct modules are touched.
        let repo_prefix = get_repo_prefix();
        let modules: HashSet<String> = paths
            .iter()
            .map(|p| get_module_for_path(p, Some(&repo_prefix)))
            .filter(|m| !m.is_empty())
            .collect();
        if modules.len() < TENTACLE_ENFORCE_MIN_MODULES {
            return None;
        }

        let mut sorted_modules: Vec<&str> = modules.iter().map(|s| s.as_str()).collect();
        sorted_modules.sort_unstable();

        Some(deny(&format!(
            "\u{1f419} TENTACLE REQUIRED: {} files across {} modules ({modules}). \
             Multi-module edits need tentacle-orchestration with a clear swarm \
             strategy, handoff trail, and explicit commit + push plan. \
             If you are the orchestrator: \
             (1) tentacle.py create <name> --scope \"<paths>\" \
             (2) tentacle.py todo <name> add \"<task>\" \
             (3) tentacle.py swarm <name> \
             (4) tentacle.py complete <name> \
             Check runtime: tentacle.py status \
             If you are a dispatched sub-agent: read bundle/manifest.json, \
             stay in declared scope, write handoff, skip git commit/push.",
            paths.len(),
            modules.len(),
            modules = sorted_modules.join(", "),
        )))
    }
}
